HARK/abstract/variables.py

```diff
@@ -0,0 +1,277 @@
+from dataclasses import dataclass, field
+from typing import Mapping, Optional, Union
+from warnings import warn
+
+import numpy as np
+import xarray as xr
+from yaml import SafeLoader, YAMLObject
+
+rng = np.random.default_rng()
+
+
+@dataclass
+class Variable(YAMLObject):
+    """
+    Abstract class for representing variables. Variables are the building blocks
+    of models. They can be parameters, states, actions, shocks, or auxiliaries.
+    """
+
+    name: str  # The name of the variable, required
+    attrs: dict = field(default_factory=dict, kw_only=True)
+    short_name: str = field(default=None, kw_only=True)
+    long_name: str = field(default=None, kw_only=True)
+    latex_repr: str = field(default=None, kw_only=True)
+    yaml_tag: str = field(default="!Variable", kw_only=False)
+    yaml_loader = SafeLoader
+
+    def __post_init__(self):
+        for key in ["long_name", "short_name", "latex_repr"]:
+            self.attrs.setdefault(key, None)
+        self.name = self.name.strip()
+        if not self.name:
+            raise ValueError("Empty variable name")
+
+    @classmethod
+    def from_yaml(cls, loader, node):
+        fields = loader.construct_mapping(node, deep=True)
+        return cls(**fields)
+
+    def __repr__(self):
+        """
+        String representation of the variable.
+
+        Returns:
+        str: The string representation of the variable.
+        """
+        return f"{self.__class__.__name__}({self.name})"
+
+
+@dataclass
+class VariableSpace(YAMLObject):
+    """
+    Abstract class for representing a collection of variables.
+    """
+
+    variables: list[Variable]
+    yaml_tag: str = field(default="!VariableSpace", kw_only=True)
+    yaml_loader = SafeLoader
+
+    def __post_init__(self):
+        """
+        Save the variables in a dictionary for easy access.
+        """
+        self.variables = {var.name: var for var in self.variables}
+
+    @classmethod
+    def from_yaml(cls, loader, node):
+        fields = loader.construct_mapping(node, deep=True)
+        return cls(**fields)
+
+
+@dataclass
+class Parameter(Variable):
+    """
+    A `Parameter` is a variable that has a fixed value.
+    """
+
+    value: Union[int, float] = 0
+    yaml_tag: str = field(default="!Parameter", kw_only=True)
+    yaml_loader = SafeLoader
+
+    def __repr__(self):
+        """
+        String representation of the parameter.
+
+        Returns:
+        str: The string representation of the parameter.
+        """
+        return f"{self.__class__.__name__}({self.name}, {self.value})"
+
+
+@dataclass
+class Parameters(VariableSpace):
+    """
+    A `Parameters` is a collection of parameters.
+    """
+
+    yaml_tag: str = "!Parameters"
+
+
+@dataclass
+class Auxiliary(Variable):
```

Comment:
I agree with Chris that 'auxiliary' is more like a macro.
I wouldn't use 'auxiliary' here, though it makes sense to have this sort of object.